### PR TITLE
[DDW-814] Discreet mode - hide sensitive information

### PR DESCRIPTION
**Pull request by renanvalentin**
This PR hides sensitive information when discreet mode is turned on. The missing features like settings page & tooltip are in a separated PR https://github.com/input-output-hk/daedalus/pull/2740

---

## Todos

- [x] Add storybook

## Screenshots

Screen Shot 2021-11-16 at 17 38 07

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1637240830173300)
- [ ] Test

---

## Test Scenarios

**Scenario 1 - Check “discreet mode” toggle**
```gherkin
Given that Daedalus is running
When I check the top bar
Then I can see an eye icon to enable the discreet mode
And by default is disabled (no strikethrough)
And upon hovering I can see a tooltip describing the action is displayed
```
**Scenario 2 - Enable “discreet mode” and check wallet balance**
```gherkin
Given that Daedalus is synced
And the discreet mode is disabled
And I have at least one synced wallet
And the wallet has transaction history
And the wallet has both ada and native tokens
When I click on the discreet mode toggle
Then I can see that the icon is now displayed with a strikethrough line
And on “Summary” tab, the following values are hidden and replaced by ***
- Total balance
- Converted balance
- Amount of ada on transactions
- Amount of tokens on transactions
- Native token amount (includes unformatted value on the tooltip)
And on “Tokens”, the following value is hidden and replaced by ***
- Native token amount
And on “Transactions” tab, the following values are hidden and replaced by ***
- Amount of ada on transactions
- Amount of tokens on transactions
And on “Send” tab, after inserting a valid address, the following values are hidden and replaced by ***
- Total balance over “ADA” input
- Native token amount on the NT dropdown (includes unformatted value on the tooltip)
And on "Settings - Wallet UTXO" the following value is hidden and replaced by ***
- Wallet balance
And on “Staking center” the following values are hidden and replaced by ***
- Balance of wallets on “Delegation center”
- Total rewards on each wallet on “Rewards”
And on Stake pools screen, the dropdown and the last sentence over the slider will be hidden 
And any value on the slider will not be changed

```
**Scenario 3 - Enable “discreet mode” and check dialogs and overlays**
```gherkin
Given that Daedalus is synced
And the discreet mode is disabled
And I have at least one synced wallet
And the wallet has both ada and native tokens
When I click on the discreet mode toggle
Then I can see that the icon is now displayed with a strikethrough line
And on step 1 on the delegation wizard, the balance of all the wallets is hidden on the dropdown and replaced by ***
And on step 1 on the register to vote wizard, the balance of all the wallets is hidden on the dropdown and replaced by ***
And on “Redeem ITN” rewards, after inserting a valid recovery phrase, the balance of all the wallets is hidden on the dropdown and replaced by ***
```
**Scenario 4 - Disable “discreet mode” and check wallet balance**
```gherkin
Given that Daedalus is synced
And the discreet mode is enabled
And I have at least one synced wallet
And the wallet has transaction history
And the wallet has both ada and native tokens
When I click on the discreet mode toggle
Then I can see that the icon shows no strikethrough
And on “Summary” tab, the following values are displayed
- Total balance
- Converted balance
- Amount of ada on transactions
- Amount of tokens on transactions
- Native token amount
And on “Tokens”, the following values are displayed
- Native token amount (includes unformatted value on the tooltip)
And on “Transactions” tab, the following values are displayed
- Amount of ada on transactions
- Amount of tokens on transactions
And on “Send” tab, after inserting a valid address, the following values are displayed
- Total balance over “ADA” input
- Native token amount on the NT dropdown (includes unformatted value on the tooltip)
And on "Settings - Wallet UTXO" the following value is displayed
- Wallet balance
And on “Staking center” the following 	the following values are displayed
- Balance of wallets on “Delegation center”
- Total rewards on each wallet on “Rewards"
And on Stake pools screen, the dropdown and the last sentence over the slider will be displayed 
And any value on the slider will not be changed

```
**Scenario 5 - Disable “discreet mode” and check dialogs and overlays**
```gherkin
Given that Daedalus is synced
And the discreet mode is enabled
And I have at least one synced wallet
And the wallet has both ada and native tokens
When I click on the discreet mode toggle
Then I can see that the icon shows no strikethrough
And on step 1 on the delegation wizard, the balance of all the wallets is displayed on the dropdown
```

**Scenario 6 -  Filter all transactions and enable discreet mode**
```gherkin
Given that Daedalus is synced
And the discreet mode is disabled
And I have at least one synced wallet with some transaction history
And I am on the “Transactions” screen
When I click on “all transactions”
And I apply the filter
And I enable the discreet mode
Then I can see that the values on “to” and “from” input are persisted
And the filter is not updated
```
**Scenario 7 -  Filter all transactions and disable discreet mode**
```gherkin
Given that Daedalus is synced
And the discreet mode is enabled
And I have at least one synced wallet with some transaction history
And I am on the “Transactions” screen
When I click on “all transactions”
And I apply the filter
And I disable the discreet mode
Then I can see that the values on “to” and “from” input are 0 and 45B respectively
And the filter is not updated
```

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Review by lucas-barros — APPROVED** [review]

File: source/renderer/app/api/utils/localStorageBridge.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// @flow
+import type { StorageKey } from '../../../../common/types/electron-store.types';
+
+export const localStorageBridge = async () => {
```
Comment (renanvalentin) [inline source/renderer/app/api/utils/localStorageBridge.js:4]: This is required to solve the following error when running storybook since `ipcRenderer` is not provided.

```
VM1721 main.c7d2b929e7e498015887.bundle.js:173935 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'on')
    at VM1721 main.c7d2b929e7e498015887.bundle.js:173935
    at new Promise (<anonymous>)
    at _callee$ (VM1721 main.c7d2b929e7e498015887.bundle.js:173918)
    at c (VM1720 vendor.dll.js:31)
    at Generator._invoke (VM1720 vendor.dll.js:31)
    at Generator.next (VM1720 vendor.dll.js:31)
    at asyncGeneratorStep (VM1721 main.c7d2b929e7e498015887.bundle.js:1206)
    at _next (VM1721 main.c7d2b929e7e498015887.bundle.js:1228)
    at VM1721 main.c7d2b929e7e498015887.bundle.js:1235
    at new Promise (<anonymous>)
```

**Comment by alexander-rukin** [comment]
image

**Comment by alexander-rukin** [comment]
image
I'd expect here instead of "< 0.01" to see just "***"

**Comment by alexander-rukin** [comment]
image

😱

**Comment by alexander-rukin** [comment]
image
lines should be aligned with the red line (or at least 1), I remember it was correct some time ago

**Comment by alexander-rukin** [comment]
image

noticed header issue - can someone from QA check if this one is new?

File: source/renderer/app/components/assets/AssetAmount.js [hunk 2]
```diff
@@ -26,47 +30,54 @@ type Props = {
...
+                amount: discreetModeFeature.hideSensitiveData(
+                  formattedTokenWalletAmount(amount, null, 0)
+                ),
+              }}
+            />
+          }
+          visible={decimals ? undefined : false}
+          className={styles.unformattedAmount}
+        >
+          <DiscreetValue ticker={{ show: Boolean(ticker), symbol: ticker }}>
```
Comment (DominikGuzei) [inline source/renderer/app/components/assets/AssetAmount.js:68]: Hm, i think we should either get rid of the `ticker` property of this component OR rename it to `CurrencyValue` and rename `ticker` to `currency` (because it could also be EUR or USD) 

its name suggests that this could be any value, not just coin ticker values so this API design does not fit IMO. Btw is it really necessary to provide the ticker value to this component? why not just wrap the number?

File: source/renderer/app/components/widgets/forms/WalletsDropdown.js [hunk 3]
```diff
@@ -33,37 +35,50 @@ export const onSearchWalletsDropdown = (
...
+  onSearch = onSearchWalletsDropdown,
+  wallets = [],
+  ...props
+}: Props) {
+  const discreetModeFeature = useDiscreetModeFeature();
+  const itemsDropdownProps = {
+    ...omit(props, ['wallets', 'options']),
+    onSearch,
   };
-  render() {
```
Comment (DominikGuzei) [inline source/renderer/app/components/widgets/forms/WalletsDropdown.js:40]: Please avoid reformatting large parts as part of a feature PR as this makes it much harder to see what actually changed.

File: source/renderer/app/features/discreet-mode/api/index.js [hunk 4]
```diff
@@ -1,14 +1,20 @@
 // @flow
-import LocalStorageApi from '../../../api/utils/localStorage';
+import { localStorageBridge } from '../../../api/utils/localStorageBridge';
 import { STORAGE_KEYS as storageKeys } from '../../../../../common/config/electron-store.config';
 
 export class DiscreetModeApi {
-  getDiscreetModeSettings = (): Promise<boolean> =>
-    LocalStorageApi.get(storageKeys.DISCREET_MODE_ENABLED, false);
+  getDiscreetModeSettings = async (): Promise<boolean> => {
+    const localStorage = await localStorageBridge();
```
Comment (DominikGuzei) [inline source/renderer/app/features/discreet-mode/api/index.js:7]: It would probably be better to pass the localstorage API as a constructor argument to `DiscreetModeApi` otherwise this is hard wired

**Comment by renanvalentin** [comment]
@alexander-rukin Pushed the fix for the tokens. Great catch!  🚀

As for the top bar @lucas-barros is investigating this behavior. The header issue on Rewards  happens when the window size is too small (Can be reproduced on develop branch)

File: source/renderer/app/features/discreet-mode/feature.js [hunk 5]
```diff
@@ -50,4 +51,20 @@ export class DiscreetMode extends Feature {
...
+
+    const content = data.split(' ');
+
+    if (content.length <= 1) {
+      return SENSITIVE_DATA_SYMBOL;
+    }
+
+    const tickerSymbol = content.pop();
+
+    return `${SENSITIVE_DATA_SYMBOL} ${tickerSymbol}`;
```
Comment (DominikGuzei) [inline source/renderer/app/features/discreet-mode/feature.js:68]: Let's make this more generic, this is 100% coupled to ticker values and even their "markup" like that the ticker symbol always is at a specific position in the string etc.

File: source/renderer/app/features/discreet-mode/ui/DiscreetValue.js [hunk 6]
```diff
@@ -3,17 +3,63 @@ import React from 'react';
...
   children: Node,
 };
 
-function DiscreetValue({ children }: Props) {
+const defaultValues = {
+  sensitiveData: {
+    className: '',
+    character: SENSITIVE_DATA_SYMBOL,
+  },
+  ticker: {
```
Comment (DominikGuzei) [inline source/renderer/app/features/discreet-mode/ui/DiscreetValue.js:27]: I don't think this component should know anything about tickers, only about the value it should cover with ***. let's move the decision to show a ticker to somewhere else (e.g the place where the discreet value is displayed)

File: source/renderer/app/features/discreet-mode/ui/DiscreetValue.story.js [hunk 7]
```diff
@@ -1,30 +1,44 @@
...
-      </>
-    </DiscreetModeFeatureProvider>
+  .addDecorator((story) => <StoryDecorator>{story()}</StoryDecorator>)
+  .addDecorator((story) => <StoryProvider>{story()}</StoryProvider>)
+  .addDecorator((story) => (
+    <DiscreetModeFeatureProvider>{story()}</DiscreetModeFeatureProvider>
+  ))
+  .add('Discreet mode disabled', () => (
+    <>
+      <DiscreetValue>123</DiscreetValue>
```
Comment (DominikGuzei) [inline source/renderer/app/features/discreet-mode/ui/DiscreetValue.story.js:35]: Why not simply `<DiscreetValue>123</DiscreetValue> ADA` instead of complicated ticker handling in the component?

**Review by DominikGuzei — CHANGES_REQUESTED** [review]
Great work! I added a few requests for improvements, especially regarding the `ticker` property on the `DiscreetValue` component which shouldn't be there IMO 😉

File: source/renderer/app/features/discreet-mode/feature.js (hunk 5 above)
Comment (renanvalentin) [reply source/renderer/app/features/discreet-mode/feature.js:68]: Removing currencyValue from `hideSensitiveData` will result in a lot of duplicated logic like:

```
let formattedValue =   formattedTokenWalletAmount(amount, null, 0)

if(discreetModeFeature.isDiscreetMode) {
       const hiddenValue = discreetModeFeature.hideSensitiveData(
                  formattedTokenWalletAmount(amount, null, 0)
                )

const currencyValue = metadata?.token || ''

formattedValue = [hiddenValue, currencyValue].join(' ').trim()
}                
```

Of course this could be moved to `formatters.js` or similar.

File: source/renderer/app/features/discreet-mode/api/index.js (hunk 4 above)
Comment (renanvalentin) [reply source/renderer/app/features/discreet-mode/api/index.js:7]: Do we really need to use dependency injection in this case? The feature is pretty small and self-contained and not leaking to other domains.

Also, what would you pass to the constructor? A the bridge ::`Promise<{get, set, etc...}>` or the instance itself? 

If the instance is used then `const [discreetModeFeature] = useState<DiscreetMode>` would have to be refactored to use `useEffect` because it has to resolve the promise first. 

I would prefer to keep like this but i'm open to suggestions of how to inject this value.

**Comment by alexander-rukin** [comment]
> image
> 
> lines should be aligned with the red line (or at least 1), I remember it was correct some time ago - [example](https://zpl.io/2pgr98r) _noticed this only because in min width with left bar opened policy id copy icon goes next line always_

image

not sure why icon needs extra line after fix 🤔

**Comment by alexander-rukin** [comment]
image

checked feature/ddw-738-implement-discreet-mode branch, still has a little overlap

**Comment by gabriela-ponce** [comment]
@alexander-rukin Regarding these issues:

1 - [Alignment of items in Native Tokens' details](https://github.com/input-output-hk/daedalus/pull/2742#issuecomment-971356621): This is displayed like that also on the released 4.4.1 version. Check [this](https://drive.google.com/file/d/1P4ZYyDfdwEmCzVjRmcZK2GgX6j3MP4kT/view?usp=sharing) screenshot.
2 - [Broken header on "Rewards" table](https://github.com/input-output-hk/daedalus/pull/2742#issuecomment-971369811): This doesn't reproduce on 4.4.1 ([screenshot](https://drive.google.com/file/d/1kaVjvIx8WZDhU-ftFKcgVOoh_VkquSu-/view?usp=sharing)), but it's on develop. Check [this](https://drive.google.com/file/d/1vEA_yxUvIkuXHFM-0nHffrdD0eh2B15B/view?usp=sharing) screenshot.

In conclusion, these are not inserted on this PR, so I'll create new tasks to address them.

File: source/renderer/app/components/widgets/forms/WalletsDropdown.js (hunk 3 above)
Comment (DominikGuzei) [reply source/renderer/app/components/widgets/forms/WalletsDropdown.js:40]: Ah ok, if the hooks refactoring was necessary no worries 👍

File: source/renderer/app/features/discreet-mode/api/index.js (hunk 4 above)
Comment (DominikGuzei) [reply source/renderer/app/features/discreet-mode/api/index.js:7]: Yeah, I think declaring the simplest interface for the required API methods is the way to go here. Regarding the eager vs lazy loading: since the electron local storage solution is already eagerly loaded + static we could replace the current `localStorageBridge` function with its contents (simple `if (isElectron)` then local storage is the electron storage, web  otherwise)

File: source/renderer/app/components/wallet/utxo/WalletUtxoDescription.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import { useDiscreetModeFeature } from '../../../features/discreet-mode';
+import type { ReactIntlMessage } from '../../../types/i18nTypes';
+
+type Props = {
+  description: ReactIntlMessage,
+  formattedWalletAmount: string,
+  walletUtxosAmount: number,
+};
+
+export function WalletUtxoDescription({
```
Comment (DominikGuzei) [inline source/renderer/app/components/wallet/utxo/WalletUtxoDescription.js:13]: i think this should be wrapped in mobx `observer`, otherwise it might not update correctly when switching the discreet mode on/off?

**Comment by miorsufianiohk** [comment]
Hi @renanvalentin , looks good on 19712.

However, there's a slight delay displaying ada amount or *** when I toggle on/off at 'Wallet UTXO Distribution'. It probably is a non issue but if the update could be synced with the other updates, it would look much better in my humble opinion. Please have a look at this [video recording](https://drive.google.com/file/d/1u6CfwpQcOHW1kCtptkJQN0i-g6OwYtwv/view?usp=sharing). Thanks.

**Comment by renanvalentin** [comment]
> Hi @renanvalentin , looks good on 19712.
> 
> However, there's a slight delay displaying ada amount or *** when I toggle on/off at 'Wallet UTXO Distribution'. It probably is a non issue but if the update could be synced with the other updates, it would look much better in my humble opinion. Please have a look at this [video recording](https://drive.google.com/file/d/1u6CfwpQcOHW1kCtptkJQN0i-g6OwYtwv/view?usp=sharing). Thanks.

@miorsufianiohk Nice catch! 👏 I just pushed a new fix for that. 

utxo

**Comment by miorsufianiohk** [comment]
Hi @renanvalentin ,

On build 19799

Thanks for fixing [this issue ](https://github.com/input-output-hk/daedalus/pull/2742#issuecomment-976519019) regarding the display on UTXO Wallets Distribution. It looks ok now 👍 

However, I found a new issue on Staking -> Stake Pools screen. When discreet mode was selected first and then an item from the dropdown list was chosen, a parsing error was displayed in the console expecting a natural number. Please see video.
[Not OK](https://drive.google.com/file/d/1pRHyEKJohbcUn_89LpsUQVj0iYKOSELD/view?usp=sharing)

Interestingly, if the mode wasn't discreet and an item from the dropdown list was chosen. Then discreet mode was chosen followed by picking an item from the dropdown list, the error wasn't displayed. Please see video.
[OK](https://drive.google.com/file/d/1eW1ScD1jSho9zQLqyhqJ9_R3HSebu_Rj/view?usp=sharing)

File: source/renderer/app/features/discreet-mode/ui/DiscreetModeFeatureInject.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import type { Node } from 'react';
+import { observer } from 'mobx-react';
+import { useDiscreetModeFeature } from '../context';
+import { DiscreetMode } from '../feature';
+
+type Props = {
+  children: (discreetModeFeature: DiscreetMode) => Node,
+};
+
+const DiscreetModeFeatureInject = ({ children }: Props) => {
```
Comment (DominikGuzei) [inline source/renderer/app/features/discreet-mode/ui/DiscreetModeFeatureInject.js:11]: Why is this necessary? I can't find any usage of this component?

File: source/renderer/app/features/discreet-mode/ui/DiscreetValue.story.js [hunk 10]
```diff
@@ -1,30 +1,44 @@
...
-      <>
-        <DiscreetValue>123</DiscreetValue>
-        <Toggle knob={boolean('Toggle', false)} />
-      </>
-    </DiscreetModeFeatureProvider>
+  .addDecorator((story) => <StoryDecorator>{story()}</StoryDecorator>)
+  .addDecorator((story) => <StoryProvider>{story()}</StoryProvider>)
+  .addDecorator((story) => (
+    <DiscreetModeFeatureProvider>{story()}</DiscreetModeFeatureProvider>
+  ))
```
Comment (DominikGuzei) [inline source/renderer/app/features/discreet-mode/ui/DiscreetValue.story.js:32]: I guess you could merge these three `addDecorator` calls into a single one & wrap the `story()` with the three provider components via the markup.

File: source/renderer/app/features/discreet-mode/ui/discreet-toggle/DiscreetToggle.js [hunk 11]
```diff
@@ -1,22 +1,23 @@
...
 import styles from './DiscreetToggle.scss';
 
-type Props = {
-  className?: string,
-  isDiscreetMode?: boolean,
-  onToggle: Function,
-};
-
-const DiscreetToggle = ({
+export const DiscreetToggleButton = ({
```
Comment (DominikGuzei) [inline source/renderer/app/features/discreet-mode/ui/discreet-toggle/DiscreetToggle.js:12]: Let's change the name to `DiscreetModeToggle` (it's not a button because it's a toggle :D) and i think the word "mode" is important here because it's not a "discreet toggle" but a toggle that switches discreet _mode_ on/off

File: storybook/stories/nodes/environment/TopBarEnvironment.stories.js [hunk 12]
```diff
@@ -15,7 +15,7 @@ import WalletTestEnvironmentLabel from '../../../../source/renderer/app/componen
 import { formattedWalletAmount } from '../../../../source/renderer/app/utils/formatters';
 import menuIconClosed from '../../../../source/renderer/app/assets/images/menu-ic.inline.svg';
 import NewsFeedIcon from '../../../../source/renderer/app/components/widgets/NewsFeedIcon';
-import DiscreetToggleTopBar from '../../../../source/renderer/app/features/discreet-mode/ui/discreet-toggle-top-bar/DiscreetToggleTopBar';
+import { DiscreetToggleTopBar } from '../../../../source/renderer/app/features';
```
Comment (DominikGuzei) [inline storybook/stories/nodes/environment/TopBarEnvironment.stories.js:18]: is this still needed?

**Review by DominikGuzei — APPROVED** [review]
Code looking great, I just found some minor issues 🎉

File: source/renderer/app/features/discreet-mode/ui/DiscreetModeFeatureInject.js (hunk 9 above)
Comment (renanvalentin) [reply source/renderer/app/features/discreet-mode/ui/DiscreetModeFeatureInject.js:11]: Removed

**Comment by nikolaglumac** [comment]
@renanvalentin can you please fix the flow issues? 🙏

**Comment by renanvalentin** [comment]
Added a new knob `toggle discreet mode` to storybook
Screen Shot 2021-11-26 at 10 50 41

**Review by gabriela-ponce — APPROVED** [review]
